pkg/emscripten/libretro-thread/libretro.js

```diff
@@ -4,237 +4,416 @@
+	arguments: ["-v", "--menu"],
+	noImageDecoding: true,
+	noAudioDecoding: true,
+
+	retroArchSend: function(msg) {
+		this.EmscriptenSendCommand(msg);
+	},
+	retroArchRecv: function() {
+		return this.EmscriptenReceiveCommandReply();
+	},
+	preRun: [
+		function(module) {
+			module.ENV["OPFS_MOUNT"] = "/home/web_user";
+		}
+	],
+	onRuntimeInitialized: function() {
+		appInitialized();
+	},
+	print: function(text) {
+		console.log("stdout:", text);
+	},
+	printErr: function(text) {
+		console.log("stderr:", text);
+	},
+	canvas: canvas,
+	totalDependencies: 0,
+	monitorRunDependencies: function(left) {
+		this.totalDependencies = Math.max(this.totalDependencies, left);
+	}
 };
 
+// read File object to an ArrayBuffer
+function readFile(file) {
+	return new Promise((resolve, reject) => {
+		let reader = new FileReader();
+		reader.onload = function() {
+			resolve(this.result);
+		}
+		reader.onerror = function(e) {
+			reject(e);
+		}
+		reader.readAsArrayBuffer(file);
+	});
+}
+
+// accept (optional) can be used to specify file extensions (string, comma delimited)
+// returns an array of {path: String, data: ArrayBuffer}
+function uploadFiles(accept) {
+	return new Promise((resolve, reject) => {
+		let input = document.createElement("input");
+		input.type = "file";
+		input.setAttribute("multiple", "");
+		if (accept) input.accept = accept;
+		input.onchange = async function() {
+			let files = [];
+			for (const file of this.files) {
+				files.push({path: file.name, data: await readFile(file)});
+			}
+			resolve(files);
+		}
+		input.oncancel = function() {
+			resolve([]);
+		}
+		input.click();
+	});
+}
 
-async function cleanupStorage()
-{
-  localStorage.clear();
-  const root = await navigator.storage.getDirectory();
-  for await (const handle of root.values()) {
-    await root.removeEntry(handle.name, {recursive: true});
-  }
-  document.getElementById("btnClean").disabled = true;
-}
-
-function appInitialized()
-{
-  /* Need to wait for the wasm runtime to load before enabling the Run button. */
-  if (retroarch_ready && filesystem_ready)
-  {
-    preLoadingComplete();
-  }
- }
-
-function preLoadingComplete() {
-   $('#icnRun').removeClass('fa-spinner').removeClass('fa-spin');
-   $('#icnRun').addClass('fa-play');
-   // Make the Preview image clickable to start RetroArch.
-   $('.webplayer-preview').addClass('loaded').click(function() {
-      startRetroArch();
-      return false;
-   });
-   $('#btnRun').removeClass('disabled').removeAttr("disabled").click(function() {
-      startRetroArch();
-      return false;
-   });
-}
-
-// Retrieve the value of the given GET parameter.
-function getParam(name) {
-   var results = new RegExp('[?&]' + name + '=([^&#]*)').exec(window.location.href);
-   if (results) {
-      return results[1] || null;
-   }
+// prompt user to upload file(s) to a dir in OPFS, e.g. "/retroarch/content"
+async function uploadFilesToDir(dir, accept) {
+	const files = await uploadFiles(accept);
```

Comment:
This is cool, I like that this gives a way to put files into the system/ folder.